File: tests/fixtures/julia_demo/billing/invoice.jl
module Billing

using CdecRules

abstract type AbstractInvoice end

struct Money
    amount::Float64
    currency::String
end

@sealed @layer "domain" struct Invoice <: AbstractInvoice
    id::String
    total::Money

    Invoice(id) = new(id, Money(0.0, "USD"))
end

mutable struct Draft
    lines::Vector{String}
end

@locked reason="agreed rounding" function formatted(inv::Invoice)::String
    return inv.id
end

@no_instantiation allow=["Money"] function summarise(inv::Invoice, verbose::Bool=false; short=true)
    m = Money(0.0, "USD")
    return inv.id
end

total_of(inv::Invoice) = inv.total

helper(x::Int) = x + 1

end # module

File: tests/fixtures/lua_demo/billing/invoice.lua
local Money = require("billing.money")

---@cdec sealed
---@cdec layer("domain")
local Invoice = {}
Invoice.__index = Invoice
Invoice.CURRENCY = "USD"

function Invoice.new(id, total)
  local self = setmetatable({}, Invoice)
  self.id = id
  self.total = total
  self._audit = {}
  return self
end

---@cdec locked(reason = "agreed rounding")
function Invoice:formatted()
  return self.id
end

---@cdec no_instantiation(allow = {"Money"})
function Invoice:summarise(verbose, ...)
  local m = Money.new(0)
  return self.id
end

local Detailed = setmetatable({}, { __index = Invoice })
Detailed.__index = Detailed

function Detailed:note() return self._note end

local function helper(x)
  return x
end

local scratch = {}

return Invoice
